### PR TITLE
Fix: missing GStaticMutex deallocations

### PR DESCRIPTION
**Pull request by dnsjts**

### CODE REVIEW

File: lib/late_ack_tracker.c [hunk 1]
```diff
@@ -198,6 +198,9 @@ void
 late_ack_tracker_free(AckTracker *s)
 {
   LateAckTracker *self = (LateAckTracker *)s;
+
+  g_static_mutex_free(&self->storage_mutex);
+
   guint32 count = ring_buffer_count(&self->ack_record_storage);
```
Comment (bazsi) [inline lib/late_ack_tracker.c]: Pls move this past the declaration in the next line.

Probably ring buffer count wouldnt use the just freed mutex, but I prefer strong visual separation of declarations vs code.

File: modules/affile/affile-dest.c [hunk 2]
```diff
@@ -318,6 +318,8 @@ affile_dw_free(LogPipe *s)
 {
   AFFileDestWriter *self = (AFFileDestWriter *) s;
   
+  g_static_mutex_free(&self->lock);
+
   log_pipe_unref((LogPipe *) self->writer);
   self->writer = NULL;
```
Comment (bazsi) [inline modules/affile/affile-dest.c]: Again its not clear whether the unref in the next line wouldn't use the mutex.

**Comment by bazsi** [comment]
I don't really understand how this got undetected so far. One possible explanation is that earlier gstaticmutezes didn't allocate memory on Linux. Maybe they do now.

But I remember fixing the dbparser leak in the last couple of weeks, maybe this is against 3.6 instead of the current master?

**Comment by lbudai** [comment]
Hmmm... I remember some lines of code in newer glibs where GStaticMutex has been deprecated.
Deprecated GStaticMutex is a simple mutex in those versions, this is why free should be called.
The problem is that if we call `g_static_mutex_free` in those cases where the user has older glib, I think it will cause problems...

So the suggested solution here is -IMHO- that we should drop GStaticMutex and go with the non-static mutexes.

@bazsi: what do you think?

**Comment by bazsi** [comment]
Hi,

It shouldn't cause problems, g_mutex_free() and g_static_mutex_free() is
not the same and the latter exists in older releases. On AIX the
GStaticMutex was allocated on-demand even in older releases, but on Linux
it was statically allocated, decreasing the number of allocations overall.
I am not sure whether these allocations are on the fast path of anything
though.

In case we decide to convert, I would get rid off the thread support from
glib and use plain pthreads or its ivykis wrappers though.

-- 
Bazsi

**Comment by bazsi** [comment]
I've found what I fixed in db-parser, there's another GStaticRWLock in patterndb.c (one layer deeper) that I remembered fixing.

More exactly this patch:

```
commit 500de3d757a7ab8c17782457833b957f665d0c90
Author: Balazs Scheidler <balazs.scheidler@balabit.com>
Date:   Tue Feb 17 15:04:36 2015 +0100

    db-parser: fixed the leaking of a GStaticRWLock
    
    This could happen at every reload and leaks a read-write mutex.
    
    Signed-off-by: Balazs Scheidler <balazs.scheidler@balabit.com>

diff --git a/modules/dbparser/patterndb.c b/modules/dbparser/patterndb.c
index c2ffb60..edb69c7 100644
--- a/modules/dbparser/patterndb.c
+++ b/modules/dbparser/patterndb.c
@@ -1859,6 +1859,7 @@ pattern_db_free(PatternDB *self)
     g_hash_table_destroy(self->state);
   if (self->timer_wheel)
     timer_wheel_free(self->timer_wheel);
+  g_static_rw_lock_free(&self->lock);
   g_free(self);
 }
```

**Comment by bazsi** [comment]
All this said, I think this should be merged, once my proposed changes are fixed, both to 3.6 and master.

@dnsjts would you fix up those or shall I?

thanks.

**Comment by dnsjts** [comment]
The proposed changes are fixed, and I will fix up those ASAP(, but please double-check the changes).

**Comment by bazsi** [comment]
:+1: this can be merged on my part to master.

**Pull request merged**